magento/magento2
# Allow only valid preferences during setup:di:compile

### Description
The `setup:di:compile` command has exclude lists to avoid loading / compiling dependency injection for 'test' classes. When preferences exist for these excluded classes, any plugins associated with the original classes are (potentially) rendered useless. This is because the child class (preference) can call the original class in a way which does not use Magento's plugin system (ie, `parent::methodName()`). And it's impossible to plugin a class that does not go through Magento's compilation process (to have interceptors created, etc). See https://github.com/magento/security-package/issues/296 for a real-world example of this problem in action.

This pull request began as a change to the regular expressions used to exclude classes, and has grown into expanding the exclude list to cover preferences. Changes include:
- Removal of invalid preferences already within the code-base.
- Alerting (exception) for preferences that do not exist.
- Alerting (exception) for preferences that do not exist because of the existing exclude lists.
- Various bug-fixes for `setup:di:compile`-related classes.

### Related Pull Requests
- https://github.com/magento/security-package/pull/297
- https://github.com/magento/inventory/pull/3315
- https://github.com/magento/magento2-page-builder/pull/743
  - Which Adobe duplicated to: https://github.com/magento-commerce/magento2-page-builder/pull/304

> Intentionally not using the magic comment which links pull requests together within the Adobe testing system, because these pull requests are related but not part of this change. Do not reinstate that magic comment.

### Fixed Issues (if relevant)
1. https://github.com/magento/security-package/issues/296
2. https://github.com/magento/magento2/issues/38517

### Manual testing scenarios
1. Create a preference for a class which does not exist. The "for" class can exist or not; the "type" class should not exist.
1. Create a preference for a class which is specifically excluded (like a "test" class). The "for" class can exist or not; the "type" class should exist, but not be eligible for dependency compilation - for example if the path includes `Test`.

Before this pull request, these would silently fail; after this pull request, an error is shown.

### Questions or comments
*None*

### Contribution checklist

### Resolved issues:
1. [x] resolves magento/magento2#38517: Allow only valid preferences during setup:di:compile
